src/dist/component/package.rs

```diff
@@ -163,19 +164,37 @@ struct MemoryBudget {
     used: usize,
 }
 
+static mut RAM_NOTICE_SHOWN: bool = false;
+
 // Probably this should live in diskio but ¯\_(ツ)_/¯
 impl MemoryBudget {
-    fn new(max_file_size: usize) -> Self {
-        const DEFAULT_UNPACK_RAM: usize = 400 * 1024 * 1024;
-        let unpack_ram = if let Ok(budget_str) = env::var("RUSTUP_UNPACK_RAM") {
-            if let Ok(budget) = budget_str.parse::<usize>() {
-                budget
-            } else {
-                DEFAULT_UNPACK_RAM
+    fn new(max_file_size: usize, effective_max_ram: usize) -> Self {
+        const DEFAULT_UNPACK_RAM_MAX: usize = 500 * 1024 * 1024;
+        let ram_for_unpacking = effective_max_ram - (100 * 1024 * 1024);
+        let default_max_unpack_ram = std::cmp::min(DEFAULT_UNPACK_RAM_MAX, ram_for_unpacking);
+        let unpack_ram = match env::var("RUSTUP_UNPACK_RAM")
+            .ok()
+            .and_then(|budget_str| budget_str.parse::<usize>().ok())
+        {
+            Some(budget) => budget,
```

Comment:
Should we do something if this is more than `effective_max_ram` ?

Reply:
Perhaps. 

One the one hand, they may know that that 100MB overhead is too generous, for instance. On the other hand they know that the autodetection code is broken for some reason. On the gripping hand we may get bug reports that 'when I tell rustup to use X RAM it breaks'.